### PR TITLE
Feature/idp integration

### PR DESCRIPTION
**Pull request by olye0216**
idp integration

### CODE REVIEW

File: src/main/java/com/icthh/xm/uaa/config/UaaConfiguration.java [hunk 1]
```diff
@@ -155,7 +163,20 @@ private TokenGranter tokenGranter(AuthorizationServerEndpointsConfigurer endpoin
             endpoints.getOAuth2RequestFactory(),
             tokenStore,
             authenticationManager);
+
+        IdpTokenGranter idpTokenGranter = new IdpTokenGranter(
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/uaa/config/UaaConfiguration.java:167]: seems IdpTokenGranter should be returned as a @Bean

File: src/main/java/com/icthh/xm/uaa/domain/UserLoginType.java [hunk 2]
```diff
@@ -35,4 +36,10 @@ public String getValue() {
         return Optional.empty();
     }
 
+    public static UserLoginType fromString(String value) {
+        return Arrays.stream(UserLoginType.values())
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/uaa/domain/UserLoginType.java:40]: Here is an inefficient enum values extraction because values() method always creates an array. 
Propose to cache enym values in static variable like:
```
    private final static Set<UserLoginType> VALUES = Set.of(UserLoginType.values());
```
and then:
```
 return VALUES.stream()
            .filter( ... 
```

File: src/main/java/com/icthh/xm/uaa/security/oauth2/idp/IdpTokenGranter.java [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    private final XmJwkTokenStore jwkTokenStore;
+    private final DomainUserDetailsService domainUserDetailsService;
+    private final TenantPropertiesService tenantPropertiesService;
+    private final UserService userService;
+    private final UserLoginService userLoginService;
+    private final GrantedAuthoritiesMapper authoritiesMapper = new NullAuthoritiesMapper();
+
+    public IdpTokenGranter(AuthorizationServerTokenServices tokenServices,
+                           ClientDetailsService clientDetailsService,
+                           @Lazy OAuth2RequestFactory requestFactory,
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/uaa/security/oauth2/idp/IdpTokenGranter.java:64]: check if the instance will be created in config method with @Bean then this @Lazy probably will not need

File: src/main/java/com/icthh/xm/uaa/security/oauth2/idp/IdpTokenGranter.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        if (userDetails.isEmpty()) {
+            log.info("User not found by identity: {}, new user will be created", userIdentity);
+            User newUser = createUser(userIdentity, idpOAuth2IdToken);
+            userDetails = Optional.of(DomainUserDetailsService.buildDomainUserDetails(userIdentity, getTenantKey(), newUser));
+        }
+        log.info("Mapped user for identity:{} is {}", userIdentity, userDetails);
+        return userDetails.get();
+    }
+
+    @LogicExtensionPoint(value = "MapIdpIdTokenToIdentity")
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/uaa/security/oauth2/idp/IdpTokenGranter.java:129]: move all LEPs to separate Service.

File: build.gradle [hunk 5]
```diff
@@ -221,6 +222,7 @@ repositories {
     maven { url 'https://oss.sonatype.org/content/repositories/releases' }
     maven { url 'https://oss.sonatype.org/content/repositories/snapshots' }
     maven { url 'https://repo.maven.apache.org/maven2' }
+    maven { url "https://oss.sonatype.org/content/repositories/snapshots/" }//TODO remove
```
Comment (vkirichenko) [inline build.gradle:225]: Do not forget to remove this

File: gradle.properties [hunk 6]
```diff
@@ -31,11 +31,11 @@ liquibase_plugin_version=2.0.1
 sonarqube_plugin_version=2.8
 spring_no_http_plugin_version=0.0.3.RELEASE
 checkstyle_version=8.24
-spotbugs_plugin_version=3.0.0
+spotbugs_plugin_version=4.2.0
 
 ## XM custom properties
 lombok_version=1.18.10
-xm_commons_version=2.1.18
+xm_commons_version=2.1.22-SNAPSHOT
```
Comment (vkirichenko) [inline gradle.properties:38]: do not forget to change commons version

File: src/main/java/com/icthh/xm/uaa/config/UaaConfiguration.java (hunk 1 above)
Comment (vkirichenko) [reply src/main/java/com/icthh/xm/uaa/config/UaaConfiguration.java:167]: After discussion we decided to leave it as it is

File: gradle.properties [hunk 7]
```diff
@@ -31,11 +31,11 @@ liquibase_plugin_version=2.0.1
 sonarqube_plugin_version=2.8
 spring_no_http_plugin_version=0.0.3.RELEASE
 checkstyle_version=8.24
-spotbugs_plugin_version=3.0.0
+spotbugs_plugin_version=4.2.0
 
 ## XM custom properties
 lombok_version=1.18.10
-xm_commons_version=2.1.18
+xm_commons_version=2.1.23-SNAPSHOT
```
Comment (amedvedchuk) [inline gradle.properties:38]: do not forget to use release commons version

**Review by amedvedchuk — APPROVED** [review]

File: build.gradle [hunk 8]
```diff
@@ -325,6 +328,7 @@ dependencies {
 
     implementation "org.springframework.security:spring-security-ldap"
     implementation "org.freemarker:freemarker:${freemarker_version}"
+    // FIXME: do we need these social related libraries?
```
Comment (vkirichenko) [inline build.gradle:331]: will be removed in separate MR

File: src/main/java/com/icthh/xm/uaa/social/package-info.java [hunk 9]
```diff
@@ -0,0 +1,5 @@
+/**
+ *
+ */
+// FIXME: to remove old social network integration
```
Comment (vkirichenko) [inline src/main/java/com/icthh/xm/uaa/social/package-info.java:4]: will be removed in separate MR

File: src/main/java/com/icthh/xm/uaa/security/DomainJwtAccessTokenDetailsPostProcessor.java [hunk 10]
```diff
@@ -17,11 +17,11 @@
 
     @LogicExtensionPoint(value = "ProcessJwtAccessTokenDetails", resolver = OptionalProfileHeaderResolver.class)
     public void processJwtAccessTokenDetails(OAuth2Authentication authentication, Map<String, Object> details) {
-        Map<String, Object> additionalDetails = new HashMap<>();
-        additionalDetails.putAll(authentication.getOAuth2Request().getRequestParameters());
+        Map<String, Object> additionalDetails = new HashMap<>(authentication.getOAuth2Request().getRequestParameters());
         // password and client_secret removed by spring
         additionalDetails.remove("grant_type");
         additionalDetails.remove("username");
+        additionalDetails.remove("token");
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/uaa/security/DomainJwtAccessTokenDetailsPostProcessor.java:24]: it's can broke existing tenant, that using token in additional deatils

File: src/main/java/com/icthh/xm/uaa/security/oauth2/idp/config/IdpConfigRepository.java [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+        return matcher.match(IDP_PUBLIC_SETTINGS_CONFIG_PATH_PATTERN, updatedKey);
+    }
+
+    @Override
+    public void onInit(String configKey, String configValue) {
+        updateIdpConfigs(configKey, configValue);
+    }
+
+    @Override
+    public void onRefresh(String updatedKey, String config) {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/uaa/security/oauth2/idp/config/IdpConfigRepository.java:101]: handle case when config was removed

File: src/main/java/com/icthh/xm/uaa/security/oauth2/idp/jwk/JwkDefinitionSource.java [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+ *
+ * @see JwkSetConverter
+ * @see JwkDefinition
+ * @see SignatureVerifier
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7517#page-10">JWK Set Format</a>
+ *
+ * @author Joe Grandja
+ * @author Michael Duergner
+ */
+public class JwkDefinitionSource {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/uaa/security/oauth2/idp/jwk/JwkDefinitionSource.java:55]: looks like library class, maybe better to add library to gradle that copy paste source code?

File: src/main/java/com/icthh/xm/uaa/security/oauth2/idp/jwk/JwkDefinition.java [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+ * This class copied from {@link org.springframework.security.oauth2.provider.token.store.jwk.JwkDefinition}
+ * and couldn't be imported cause of package private access.
+ * The base representation of a JSON Web Key (JWK).
+ *
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7517">JSON Web Key (JWK)</a>
+ *
+ * @author Joe Grandja
+ * @author Michael Duergner
+ */
+public abstract class JwkDefinition {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/uaa/security/oauth2/idp/jwk/JwkDefinition.java:28]: looks like library class, maybe better to add library to gradle that copy paste source code?

File: src/main/java/com/icthh/xm/uaa/security/oauth2/idp/jwk/JwkAttributes.java [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+ * <p>
+ * Reason: we need custom implementation of JwkDefinitionSource class
+ * which impossible to import and override - it has package private access.
+ * <p>
+ * This class is required for {@link XmJwkVerifyingJwtAccessTokenConverter} implementation.
+ *
+ * @author Joe Grandja
+ * @author Michael Duergner
+ */
+public final class JwkAttributes {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/uaa/security/oauth2/idp/jwk/JwkAttributes.java:35]: looks like library class, maybe better to add library to gradle that copy paste source code?

File: src/main/java/com/icthh/xm/uaa/security/oauth2/idp/jwk/EllipticCurveJwkDefinition.java [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+ * This class copied from {@link org.springframework.security.oauth2.provider.token.store.jwk.EllipticCurveJwkDefinition}.
+ * and couldn't be imported cause of package private access.
+ * A JSON Web Key (JWK) representation of an Elliptic Curve key.
+ *
+ * @author Michael Duergner <michael@sprucehill.io>
+ * @since 2.3
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7517">JSON Web Key (JWK)</a>
+ * @see <a target="_blank" href="https://tools.ietf.org/html/rfc7518#page-28">JSON Web Algorithms (JWA)</a>
+ */
+public final class EllipticCurveJwkDefinition extends JwkDefinition {
```
Comment (sergeysenja1992) [inline src/main/java/com/icthh/xm/uaa/security/oauth2/idp/jwk/EllipticCurveJwkDefinition.java:28]: why we copy paste source code on spring-security-oauth2 ?

how about migrations, bug fixes etc?

File: src/main/java/com/icthh/xm/uaa/security/DomainTokenServices.java [hunk 16]
```diff
@@ -48,6 +49,8 @@
     @Setter
     private TokenStore tokenStore;
     @Setter
+    private XmJwkTokenStore jwkTokenStore;
```
Comment (amedvedchuk) [inline src/main/java/com/icthh/xm/uaa/security/DomainTokenServices.java:52]: @olye0216 @vkirichenko why do we need XmJwkTokenStore here?

File: src/main/java/com/icthh/xm/uaa/security/DomainJwtAccessTokenDetailsPostProcessor.java (hunk 10 above)
Comment (olye0216) [reply src/main/java/com/icthh/xm/uaa/security/DomainJwtAccessTokenDetailsPostProcessor.java:24]: Remark applied: removed.

File: src/main/java/com/icthh/xm/uaa/security/oauth2/idp/config/IdpConfigRepository.java (hunk 11 above)
Comment (olye0216) [reply src/main/java/com/icthh/xm/uaa/security/oauth2/idp/config/IdpConfigRepository.java:101]: handled in 
`if (isRawClientsConfigurationEmpty && isValidClientsConfigurationEmpty) {
                log.warn("For tenant [{}] provided IDP public client configs not present." +
                    "Removing client configs and claim verifiers from storage", tenantKey);
                idpClientConfigs.remove(tenantKey);
                jwtClaimsSetVerifiersHolder.remove(tenantKey);
                return;
            }`

File: src/main/java/com/icthh/xm/uaa/security/oauth2/idp/jwk/JwkDefinitionSource.java (hunk 12 above)
Comment (olye0216) [reply src/main/java/com/icthh/xm/uaa/security/oauth2/idp/jwk/JwkDefinitionSource.java:55]: It won't solve problems with package private access in the class definition. 
Also, some methods, ex. JwkDefinitionSource#getDefinitionLoadIfNecessary should have public access to provide the ability to override their behaiviour.
Also, JwkDefinitionSource should have a default constructor for correct work of custom XmJwkDefinitionSource. which is missing in the original class.

File: src/main/java/com/icthh/xm/uaa/security/oauth2/idp/jwk/JwkDefinition.java (hunk 13 above)
Comment (olye0216) [reply src/main/java/com/icthh/xm/uaa/security/oauth2/idp/jwk/JwkDefinition.java:28]: It won't solve problems with package private access in the class and its inner classes definition.

File: src/main/java/com/icthh/xm/uaa/security/oauth2/idp/jwk/EllipticCurveJwkDefinition.java (hunk 15 above)
Comment (olye0216) [reply src/main/java/com/icthh/xm/uaa/security/oauth2/idp/jwk/EllipticCurveJwkDefinition.java:28]: We copy source code from the spring-security-oauth2 package only for one reason - we have demands to customize the behavior of some methods, that involved in the token validation chain and cause of original spring-security-oauth2 has package-private access class/methods within it is impossible to customize them.

File: src/main/java/com/icthh/xm/uaa/security/DomainTokenServices.java (hunk 16 above)
Comment (olye0216) [reply src/main/java/com/icthh/xm/uaa/security/DomainTokenServices.java:52]: Currently, there are no needs for XmJwkTokenStore in DomainTokenService, but it placed there for further implementations for stateful mode. It's some kind of "bread crumbs" and can be removed if necessary.

Comment (olye0216) [reply src/main/java/com/icthh/xm/uaa/security/DomainTokenServices.java:52]: XmJwkTokenStore deleted in this class as unused.

File: src/main/java/com/icthh/xm/uaa/domain/UserLoginType.java [hunk 17]
```diff
@@ -35,4 +38,12 @@ public String getValue() {
         return Optional.empty();
     }
 
+    public static UserLoginType fromString(String value) {
```
[the remark this answers is not in the document]
Comment (olye0216) [reply src/main/java/com/icthh/xm/uaa/domain/UserLoginType.java:41]: standard method valueOf() searches for exact value match, so it could lead to errors during runtime.
For that reason custom implementation used.

**Review by vkirichenko — APPROVED** [review]
